sql/plan/indexed_join.go

```diff
@@ -0,0 +1,253 @@
+	}), nil
+}
+
+// indexedJoinIter is an iterator that iterates over every row in the primary table and performs an index lookup in
+// the secondary table for each value
+type indexedJoinIter struct {
+	primary              sql.RowIter
+	primaryRow           sql.Row
+	index                sql.Index
+	secondaryIndexAccess *IndexedTableAccess
+	secondaryProvider    sql.Node
+	secondary            sql.RowIter
+	primaryTableExpr     sql.Expression
+	cond                 sql.Expression
+	joinType             JoinType
+
+	ctx        *sql.Context
+	foundMatch bool
+	rowSize    int
+}
+
+func (i *indexedJoinIter) loadPrimary() error {
+	if i.primaryRow == nil {
+		r, err := i.primary.Next()
+		if err != nil {
+			return err
+		}
+
+		i.primaryRow = r
+		i.foundMatch = false
+	}
+
+	return nil
+}
+
+func (i *indexedJoinIter) loadSecondary() (sql.Row, error) {
+	if i.secondary == nil {
+		// evaluate the primary row against the primary table expression to get the secondary table lookup key
+		key, err := i.primaryTableExpr.Eval(i.ctx, i.primaryRow)
+		if err != nil {
+			return nil, err
+		}
+
+		lookup, err := i.index.Get(key)
+		if err != nil {
+			return nil, err
+		}
+
+		err = i.secondaryIndexAccess.SetIndexLookup(i.ctx, lookup)
+		if err != nil {
+			return nil, err
+		}
+
+		span, ctx := i.ctx.Span("plan.IndexedJoin indexed lookup")
+		rowIter, err := i.secondaryProvider.RowIter(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		i.secondary = sql.NewSpanIter(span, rowIter)
+	}
+
+	secondaryRow, err := i.secondary.Next()
+	if err != nil {
+		if err == io.EOF {
+			i.secondary = nil
+			i.primaryRow = nil
+			return nil, io.EOF
+		}
+		return nil, err
+	}
+
+	return secondaryRow, nil
+}
+
+func (i *indexedJoinIter) Next() (sql.Row, error) {
+	for {
+		if err := i.loadPrimary(); err != nil {
+			return nil, err
+		}
+
+		primary := i.primaryRow
+		secondary, err := i.loadSecondary()
+		if err != nil {
+			if err == io.EOF {
+				if !i.foundMatch && (i.joinType == JoinTypeLeft || i.joinType == JoinTypeRight) {
+					return i.buildRow(primary, nil), nil
+				}
+				continue
+			}
+			return nil, err
+		}
+
+		row := i.buildRow(primary, secondary)
+		matches, err := conditionIsTrue(i.ctx, row, i.cond)
+		if err != nil {
+			return nil, err
+		}
+
+		if !matches {
+			continue
+		}
+
+		i.foundMatch = true
+		return row, nil
+	}
+}
+
+func conditionIsTrue(ctx *sql.Context, row sql.Row, cond sql.Expression) (bool, error) {
+	v, err := cond.Eval(ctx, row)
+	if err != nil {
+		return false, err
+	}
+
+	// Expressions containing nil evaluate to nil, not false
+	return v == true, nil
+}
+
+// buildRow builds the result set row using the rows from the primary and secondary tables
+func (i *indexedJoinIter) buildRow(primary, secondary sql.Row) sql.Row {
+	row := make(sql.Row, i.rowSize)
+
+	copy(row, primary)
+	copy(row[len(primary):], secondary)
+
+	return row
+}
+
+func (i *indexedJoinIter) Close() (err error) {
+	i.secondary = nil
```

Comment:
This looks wrong given `i.secondary` handling below.

Reply:
Good catch.